### PR TITLE
test(24.04): add LXD backend to Spread

### PR DESCRIPTION
**Pull request by cjdcordeiro**
(same as #316 , for 24.04)

### CODE REVIEW

**Comment by cjdcordeiro** [comment]
@Meulengracht can you please check if this fixes your privileged tests? (use `spread lxd`)

**Comment by Meulengracht** [comment]
> @Meulengracht can you please check if this fixes your privileged tests? (use `spread lxd`)

lxd fails to come up, it's stuck doing `lxc exec ubuntu-noble -- pgrep sshd`

**Comment by Meulengracht** [comment]
> @Meulengracht but can you even launch an lxd instance outside spread? If you do `lxd launch ubuntu:24.04 u1`, will it come up?

Yea it works just fine - I use lxd for multiple purposes

**Comment by cjdcordeiro** [comment]
And when you say it's stuck, does it timeout? Note that LXD will first pull/refresh the image, which could take a while depending on your connection. 

If you do the `lxc launch...` I mentioned above and then `spread lxd`, will it still get stuck?

**Review by linostar — APPROVED** [review]
LGTM, thanks!

**Comment by Meulengracht** [comment]
> And when you say it's stuck, does it timeout? Note that LXD will first pull/refresh the image, which could take a while depending on your connection.
> 
> If you do the `lxc launch...` I mentioned above and then `spread lxd`, will it still get stuck?

Sorry for the incredibly slow reply here - So I looked a bit more into this, it seems that `sshd` does not come up, it tells me the service unit does not exist. The container is running fine, and I can `shell` into it.

File: spread.yaml [hunk 1]
```diff
@@ -11,6 +11,35 @@ exclude:
...
+      # Ideally, we would add the ubuntu-minimal remote
+      # e.g. https://cloud-images.ubuntu.com/minimal/releases/jammy/
+      # but that would effectively change the host's LXC configurations.
+
+      echo "Allocating $SPREAD_SYSTEM..."
+      lxc launch --ephemeral ubuntu:$release $SPREAD_SYSTEM
+      until lxc exec $SPREAD_SYSTEM -- pgrep sshd &>/dev/null
+      do
+        sleep 5
+      done
```
Comment (Meulengracht) [inline spread.yaml:31]:
```suggestion
      until lxc exec $SPREAD_SYSTEM -- systemctl status | grep "running"
      do
        sleep 5
      done
      lxc exec $SPREAD_SYSTEM -- systemctl enable --now ssh
```

Comment (Meulengracht) [reply spread.yaml:31]: This made it work for me

**Review by Meulengracht — COMMENTED** [review]
Added a suggestion for improvement. That fixed it for me

File: spread.yaml (hunk 1 above)
Comment (cjdcordeiro) [reply spread.yaml:31]: Hum, that seems redundant but harmless...giving it a try

File: spread.yaml [hunk 2]
```diff
@@ -11,6 +11,36 @@ exclude:
...
+      until lxc exec $SPREAD_SYSTEM -- systemctl status | grep "running"
+      do
+        sleep 5
+      done
+      lxc exec $SPREAD_SYSTEM -- systemctl enable --now ssh
+      lxc exec $SPREAD_SYSTEM -- sed -i 's/^\s*#\?\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' /etc/ssh/sshd_config
+      lxc exec $SPREAD_SYSTEM -- bash -c "sed -i 's/^\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/# COMMENTED OUT BY SPREAD: \0/' /etc/ssh/sshd_config.d/* || true"
+      lxc exec $SPREAD_SYSTEM -- bash -c "test -d /etc/ssh/sshd_config.d && echo -e 'PermitRootLogin=yes\nPasswordAuthentication=yes' > /etc/ssh/sshd_config.d/00-spread.conf"
+      lxc exec $SPREAD_SYSTEM -- bash -c "echo root:${SPREAD_SYSTEM_PASSWORD} | chpasswd"
+      lxc exec $SPREAD_SYSTEM -- killall -HUP sshd
```
Comment (Meulengracht) [inline spread.yaml:37]: Seems more appropriate?
```suggestion
      lxc exec $SPREAD_SYSTEM -- systemctl restart ssh
```

Comment (cjdcordeiro) [reply spread.yaml:37]: here I'm actually reproducing the official spread backend behaviour: https://github.com/snapcore/spread/blob/ded9133cdbceaf01f8a1c9decf6ff9ea56e194d6/spread/lxd.go#L486

So I'd rather keep it that way

**Review by rebornplusplus — APPROVED** [review]
Looks good to me! Hope it works! :crossed_fingers:

**Pull request merged**